server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -159,32 +158,37 @@ public SearchPhaseController.ReducedQueryPhase reduce() throws Exception {
         if (f != null) {
             throw f;
         }
-
+        List<QuerySearchResult> buffer;
+        synchronized (this) {
+            buffer = this.buffer;
+            buffer = buffer == null ? Collections.emptyList() : buffer;
+            this.buffer = null;
+        }
         // ensure consistent ordering
-        sortBuffer();
+        buffer.sort(RESULT_COMPARATOR);
         final TopDocsStats topDocsStats = this.topDocsStats;
+        var mergeResult = this.mergeResult;
```

Comment:
This question is probably because of my ignorance in this area but, why are we not doing this inside the synchronize block? it feels natural to collect all the stuff we are going to merge there.

Reply:
There's no functional difference doing it in or after the sync block. Either way we're not ever going to contend on anything because once we hit `reduce` no more results are expected to come in.
It's really just an aesthetic choice almost. My thinking was just that I'd like to have all the non-volatile store/load memory effects dealt with first to create a situation I can reason about well and then swap this one out asap.
Doing it under the mutex won't change anything unless I make the field non-volatile and we shouldn't do unnecessary work under the mutex (I merely felt I couldn't get away with `synchronize(this) {}` in a review though it would be good enough :D). Maybe we could look into follow-ups that rely more on the sync block though :)?

Reply:
sounds good to me.